lib/SPIRV/SPIRVOps.cpp

```diff
@@ -740,6 +740,102 @@ static LogicalResult verify(spirv::VariableOp varOp) {
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// spv.CompositeExtractOp
+//===----------------------------------------------------------------------===//
+
+// Extracts the Result Type from Composite Type iterating all over the indices,
+// with check for out of bounds.
+static Type getElementTypeByIndices(OpAsmParser *parser, OperationState *state,
```

Comment:
I've added CompositeType as a base class between ArrayType, StructType. The CompositeTypeStorage does not have a static "construct" method, because we don't need to directly construct it and therefore it does not have a operator == for the hash function.
It provides a method getElementTypByIndices which consumes array of int32_t, as result it returns a pair<Type, StringRef> on success it would be a type selected by the last provided index, on failure a nullpr with an error message. I didn't come with better idea, how to handle different error messages and don't insert  asserts.